### DISCUSSION TITLE
Use exported website without local server

### DISCUSSION THREAD

**Discussion by mkozachok** [vercel/next.js#40479] [Help]
My client requests the feature to have a copy of the website (multipage) and be able to use on a computer as static website without local server. Why? Clients are not developers, so settling server locally is too hard for them.

```javascript
// next.config.js
module.exports = {
    trailingSlash: true,
    assetPrefix: ".",
}
```

I have used `next export` and have issue with images. I replaced `Image` with `img` and exported the website successfully. 
And it works with `npx static-server`.

It also works when you just open "path-to-project/out/index.html".

But navigation does not works. This is example, when you want go to "/works".
![image](https://user-images.githubusercontent.com/10548300/189646519-d778f243-75f3-44e6-afb6-041ac5ad6a82.png)

Ok, lets enter  "path-to-project/out/works/index.html". And we will see the blank page, because js- and css- files are not loaded.
![image](https://user-images.githubusercontent.com/10548300/189647377-af756d73-8a88-4f71-bbb1-f1d1e3ca8c88.png)

This is because it tries to load files from folder "out/works", but files are "one folder up".

**How to deal with issues?** 
Maybe it is a better way to achieve the goal?

**Comment by icyJoseph** [comment]
Hi,

The History API doesn't work if the protocol is `file`. It ought to be `http` or `https`.

That is to say that any SPA will not work in `file` protocol, even if you use `react-router-dom`, so this is not isolated to Next.js -

> Personally I think https://www.npmjs.com/package/serve is pretty good to serve static Next.js apps

Rather I think you could do, given zero technical knowledge/ability from your client, is to package the app as an actual native executable, with something as simple as https://github.com/tauri-apps/tauri - Tauri takes HTML, CSS and JS as inputs and it wraps it all into a native application. Then your users just need to click the executable and voíla!

**Reply by mkozachok** [reply]
- `serve` is good option but for developers (you need node to be installed and to know how run terminal)
-  Yes, `tauri` is a good option. It takes more time for developing. But maybe it is better to choose electron?

**Reply by icyJoseph** [reply]
Up to you. For me tauri would be good because it's a step afterwards, but maybe Rust is not your thing. At least you know the issue is using the history API over to the file protocol.

You could also just use plain anchor tags, and window.location, to give up the SPA experience entirely. Again, up to you.

**Reply by mkozachok** [reply]
> plain anchor tags,

yes it is good work around, and I have tried it. it works.

> Ok, lets enter "path-to-project/out/works/index.html". And we will see the blank page, because js- and css- files are not loaded.

for now, I do not know how to solve issue with js file loading in deep folders.

**Reply by mkozachok** [reply]
In some way it needs to rewrite src as

- out/index.html // <script src="./_next/....">
- - out/works/index.html // <script src="../_next/....">
- - - out/works/my-work/index.html // <script src="../../_next/....">

but how to do this?

**Reply by icyJoseph** [reply]
Ah wait no, the scripts can't do it, and for some reason your images have `///` as path start. Looks like the images are having an entirely different issue. 

Why not host the APP for your client somewhere? Put it behind basic auth and that's it - I reckon there's a need for the app to be on their computers, or just use Proton, Electron, Tauri, etc. 

For the record, long ago, I run into the file protocol vs History API problem, because some users right click saved the Create React App, we had shipped. That was not fun at all.

**Reply by mkozachok** [reply]
Is there some way to know that it is doing build for export?

```javascript
// some component with image
if (export) {
return <img />
} else {
return <Image/>
}
```
